### PR TITLE
1955 temperature sensor sync for temperature logs and breaches

### PR DESCRIPTION
**Pull request by adrianwb**
Fixes part of #1955

# 👩🏻‍💻 What does this PR do? 
Add in sync for `temperature_log`, `temperature_breach` and `temperature_breach_config`

# 🧪 How has/should this change been tested? 
Just the automated tests, like any other sync table

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

**Comment by adrianwb** [comment]
It's all compiling up ok but `cargo test sync_pull` is failing (everything else passes):

```
running 1 test
test sync::test::pull_and_push::test_sync_pull_and_push ... FAILED

failures:

---- sync::test::pull_and_push::test_sync_pull_and_push stdout ----
thread 'sync::test::pull_and_push::test_sync_pull_and_push' panicked at 'TemperatureBreachConfig row not found: 997812e0c33911eb9757779d39ae2dbd', service\src\sync\test\mod.rs:187:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and I can't figure out why this isn't working when the very similar `TemperatureBreach`  & `TemperatureLog` added at the same time both seem to be passing - I've looked at the `test_sync_pull_and_push.sqlite` DB in the `open-msupply\server\service\test_output` folder and there is indeed no `temperature_breach_config` record => I'm guessing there's something which prevents it being created in the first place??

**Review by andreievg — APPROVED** [review]
Thanks looks great, one change request though (but approving anyways, i'll be away till next tuesday, so good to get as much as possible in before you go on holiday, btw I would like to work on some things while you away if that's ok, so if you have any carry over things, can you please put in branch)

> It's all compiling up ok but cargo test sync_pull is failing (everything else passes):

It's due to this one: 

```diff
diff --git a/server/service/src/sync/translations/mod.rs b/server/service/src/sync/translations/mod.rs
index e9342531c..a453da43f 100644
--- a/server/service/src/sync/translations/mod.rs
+++ b/server/service/src/sync/translations/mod.rs
@@ -80,6 +80,7 @@ pub(crate) fn all_translators() -> SyncTranslators {
         Box::new(sensor::SensorTranslation {}),
         Box::new(temperature_log::TemperatureLogTranslation {}),
         Box::new(temperature_breach::TemperatureBreachTranslation {}),
+        Box::new(temperature_breach_config::TemperatureBreachConfigTranslation {}),
         Box::new(clinician::ClinicianTranslation {}),
         Box::new(clinician_store_join::ClinicianStoreJoinTranslation {}),
         // Remote-Central (site specific)
diff --git a/server/service/src/sync/translations/temperature_breach_config.rs b/server/service/src/sync/translations/temperature_breach_config.rs
index 0f1830782..a7a850bc3 100644
--- a/server/service/src/sync/translations/temperature_breach_config.rs
+++ b/server/service/src/sync/translations/temperature_breach_config.rs
@@ -43,7 +43,7 @@ pub(crate) struct TemperatureBreachConfigTranslation {}
 impl SyncTranslation for TemperatureBreachConfigTranslation {
     fn pull_dependencies(&self) -> PullDependency {
         PullDependency {
-            table: LegacyTableName::TEMPERATURE_BREACH,
+            table: LEGACY_TABLE_NAME,
             dependencies: vec![LegacyTableName::STORE],
         }
     }

```

with `CodeLLDB` in vs code and rust analyzer, you can get this inlay hints, to debug and run tests or main etc, and breakpoints work. 

Screenshot 2023-08-30 at 11 52 58 PM

I first tried to [remove comment one this one](https://github.com/openmsupply/open-msupply/blob/bf2b50170eefd1f623405166b0046743beb21873/server/service/src/sync/test/pull_and_push.rs#L28) and run the test (through the inlay hint, otherwise you would need `--nocapture` flag for cargo to see output of tests). That didn't show that translator was not found, so put breakpoint in translator, nothing hit, so [put conditional breakpoint here](https://github.com/openmsupply/open-msupply/blob/bf2b50170eefd1f623405166b0046743beb21873/server/service/src/sync/translation_and_integration.rs#L45): 

```
sync_record.record_id == "997812e0c33911eb9757779d39ae2dbd"
```

nothing hit again, so realised that the records is not even being grabbed from sync_buffer.

This needs improving at some point, it's pretty easy to miss things, will look at improving at some point. I am very surprised that there are no compilation errors, since pub(crate) struct is not actually used anywhere but tests

**Comment by andreievg** [comment]
We can do this to let compiler help:

```diff
diff --git a/server/service/src/sync/translations/mod.rs b/server/service/src/sync/translations/mod.rs
index e9342531c..89f35788c 100644
--- a/server/service/src/sync/translations/mod.rs
+++ b/server/service/src/sync/translations/mod.rs
@@ -81,6 +81,8 @@ pub(crate) fn all_translators() -> SyncTranslators {
         Box::new(temperature_log::TemperatureLogTranslation {}),
         Box::new(temperature_breach::TemperatureBreachTranslation {}),
         Box::new(clinician::ClinicianTranslation {}),
+        // if this is missing there will be compilation error
+        temperature_breach_config::boxed(),
         Box::new(clinician_store_join::ClinicianStoreJoinTranslation {}),
         // Remote-Central (site specific)
         Box::new(name_store_join::NameStoreJoinTranslation {}),
diff --git a/server/service/src/sync/translations/temperature_breach_config.rs b/server/service/src/sync/translations/temperature_breach_config.rs
index 0f1830782..25ac389d2 100644
--- a/server/service/src/sync/translations/temperature_breach_config.rs
+++ b/server/service/src/sync/translations/temperature_breach_config.rs
@@ -39,7 +39,13 @@ pub struct LegacyTemperatureBreachConfigRow {
     pub maximum_temperature: f64,
 }
 
-pub(crate) struct TemperatureBreachConfigTranslation {}
+// Needs to be added to all_translators()
+#[deny(dead_code)]
+pub(crate) fn boxed() -> Box<dyn SyncTranslation> {
+    Box::new(TemperatureBreachConfigTranslation {})
+}
+
+pub(super) struct TemperatureBreachConfigTranslation {}
 impl SyncTranslation for TemperatureBreachConfigTranslation {
     fn pull_dependencies(&self) -> PullDependency {
         PullDependency {
```

And the translation trait can be made more usable, the translator needs to implement a method that returns a table name (the translator trait has a default implementation to check that table, this can be overridden, but it means we don't need [to do this ones](https://github.com/openmsupply/open-msupply/blob/bf2b50170eefd1f623405166b0046743beb21873/server/service/src/sync/translations/temperature_breach_config.rs#L56-L58), and also don't have to specify own table in dependency)

File: server/repository/src/migrations/v1_02_00/sensor.rs [hunk 1]
```diff
@@ -27,6 +62,18 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
                 CREATE TRIGGER sensor_trigger
                 AFTER INSERT OR UPDATE OR DELETE ON sensor
                 FOR EACH ROW EXECUTE PROCEDURE update_changelog();
+
+                CREATE TRIGGER temperature_breach_trigger
+                AFTER INSERT OR UPDATE OR DELETE ON temperature_breach
+                FOR EACH ROW EXECUTE PROCEDURE update_changelog();
+
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/repository/src/migrations/v1_02_00/sensor.rs:68]: Done here: https://github.com/openmsupply/open-msupply/pull/2152/commits/bf73a5c4c87722c96ffcc9b0df9f26ec97e7262a

File: server/service/src/sync/translations/temperature_breach.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+fn match_pull_table(sync_record: &SyncBufferRow) -> bool {
+    sync_record.table_name == LEGACY_TABLE_NAME
+}
+fn match_push_table(changelog: &ChangelogRow) -> bool {
+    changelog.table_name == ChangelogTableName::TemperatureBreach
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub enum LegacyTemperatureBreachType {
```
[the remark this answers is not in the document]
Comment (adrianwb) [reply server/service/src/sync/translations/temperature_breach.rs:28]: Done here: https://github.com/openmsupply/open-msupply/pull/2152/commits/bf73a5c4c87722c96ffcc9b0df9f26ec97e7262a

**Comment by adrianwb** [comment]
> Thanks looks great, one change request though (but approving anyways, i'll be away till next tuesday, so good to get as much as possible in before you go on holiday, btw I would like to work on some things while you away if that's ok, so if you have any carry over things, can you please put in branch)

I didn't get as far as I'd hoped but by all means, please carry on and make whatever changes are needed. I'll commit any work in progress.

> > It's all compiling up ok but cargo test sync_pull is failing (everything else passes):
> 
> It's due to this one:
> 
> ```diff
> diff --git a/server/service/src/sync/translations/mod.rs b/server/service/src/sync/translations/mod.rs
> index e9342531c..a453da43f 100644
> --- a/server/service/src/sync/translations/mod.rs
> +++ b/server/service/src/sync/translations/mod.rs
> @@ -80,6 +80,7 @@ pub(crate) fn all_translators() -> SyncTranslators {
>          Box::new(sensor::SensorTranslation {}),
>          Box::new(temperature_log::TemperatureLogTranslation {}),
>          Box::new(temperature_breach::TemperatureBreachTranslation {}),
> +        Box::new(temperature_breach_config::TemperatureBreachConfigTranslation {}),
>          Box::new(clinician::ClinicianTranslation {}),
>          Box::new(clinician_store_join::ClinicianStoreJoinTranslation {}),
>          // Remote-Central (site specific)
> diff --git a/server/service/src/sync/translations/temperature_breach_config.rs b/server/service/src/sync/translations/temperature_breach_config.rs
> index 0f1830782..a7a850bc3 100644
> --- a/server/service/src/sync/translations/temperature_breach_config.rs
> +++ b/server/service/src/sync/translations/temperature_breach_config.rs
> @@ -43,7 +43,7 @@ pub(crate) struct TemperatureBreachConfigTranslation {}
>  impl SyncTranslation for TemperatureBreachConfigTranslation {
>      fn pull_dependencies(&self) -> PullDependency {
>          PullDependency {
> -            table: LegacyTableName::TEMPERATURE_BREACH,
> +            table: LEGACY_TABLE_NAME,
>              dependencies: vec![LegacyTableName::STORE],
>          }
>      }
> ```

Thanks - it's obvious now that you've pointed it out! Fixed here: https://github.com/openmsupply/open-msupply/pull/2152/commits/bf73a5c4c87722c96ffcc9b0df9f26ec97e7262a

> with `CodeLLDB` in vs code and rust analyzer, you can get this inlay hints, to debug and run tests or main etc, and breakpoints work.
> 
> Screenshot 2023-08-30 at 11 52 58 PM
> 
> I first tried to [remove comment one this one](https://github.com/openmsupply/open-msupply/blob/bf2b50170eefd1f623405166b0046743beb21873/server/service/src/sync/test/pull_and_push.rs#L28) and run the test (through the inlay hint, otherwise you would need `--nocapture` flag for cargo to see output of tests). That didn't show that translator was not found, so put breakpoint in translator, nothing hit, so [put conditional breakpoint here](https://github.com/openmsupply/open-msupply/blob/bf2b50170eefd1f623405166b0046743beb21873/server/service/src/sync/translation_and_integration.rs#L45):
> 
> ```
> sync_record.record_id == "997812e0c33911eb9757779d39ae2dbd"
> ```
> 
> nothing hit again, so realised that the records is not even being grabbed from sync_buffer.
> 
> This needs improving at some point, it's pretty easy to miss things, will look at improving at some point. I am very surprised that there are no compilation errors, since pub(crate) struct is not actually used anywhere but tests

Thanks for all that too - still haven't used the debugger much so that's helpful advice :-)

**Comment by andreievg** [comment]
Thanks @adrianwb, re-tested (with sqlite) with current state of the branch, and merging

**Pull request merged**